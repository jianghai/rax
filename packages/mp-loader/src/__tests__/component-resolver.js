const { join } = require('path');
const resolveDependencyComponents = require('../component-resolver');

const { resolveComponentPath } = resolveDependencyComponents;

const mockPwd = '/@mock-path@/';
const projectPath = join(mockPwd, 'test-project'); // mocked path
const pagePath = join(projectPath, 'pages/foo/foo.js');
const componentPath = join(projectPath, '/components/foo/foo');
const npmPkgPath = join(projectPath, 'node_modules/npm-pkg');
const subComponentPath = join(projectPath, 'pages/foo/sub-component/index');
const pluginComponentPath = 'plugin://foo-plugin/foo-component';

describe('component resolver', () => {
  const r = resolveComponentPath;
  it('should resolve absolute component path', () => {
    expect(
      r('/components/foo/foo', projectPath, pagePath)
    ).toEqual(componentPath);
  });

  it('should resolve relative compoennt path', () => {
    expect(
      r('../../components/foo/foo', projectPath, pagePath)
    ).toEqual(componentPath);

    expect(
      r('./sub-component/index', projectPath, pagePath)
    ).toEqual(subComponentPath);
  });

  it('should resolve npm component', () => {
    expect(
      r('npm-pkg', projectPath, pagePath)
    ).toEqual(npmPkgPath);
  });

  it('should support plugin component', () => {
    expect(
      r(pluginComponentPath, projectPath, pagePath)
    ).toEqual(pluginComponentPath);
  });

  it('should resolve config properly', () => {
    const usingComponents = {
      npm: 'npm-pkg/es/card/index',
      plugin: pluginComponentPath,
      relative: '../../components/foo/index',
      absolute: '/components/foo/index',
      self: './foo.js',
    };
    expect(
      resolveDependencyComponents({ usingComponents }, projectPath, pagePath)
    ).toMatchSnapshot();
  });
});
