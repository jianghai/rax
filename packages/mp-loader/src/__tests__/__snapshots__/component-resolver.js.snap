// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component resolver should resolve config properly 1`] = `
Object {
  "absolute": "/@mock-path@/test-project/components/foo/index",
  "npm": "/@mock-path@/test-project/node_modules/npm-pkg/es/card/index",
  "plugin": "plugin://foo-plugin/foo-component",
  "relative": "/@mock-path@/test-project/components/foo/index",
  "self": "__SELF__",
}
`;
